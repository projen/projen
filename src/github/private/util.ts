export function secretToString(secretName: string): string {
  return `\${{ secrets.${secretName} }}`;
}

export function context(value: string) {
  return `\${{ ${value} }}`;
}

// Checks if part of the file path is hidden
export function isHiddenPath(path: string) {
  return /(^|\/)\.[^\/\.]/g.test(path);
}

// Helper to assert a path is not hidden
export function ensureNotHiddenPath(value: string, name: string) {
  if (isHiddenPath(value)) {
    throw Error(`${name} cannot be a hidden path, got: ${value}`);
  }
}
