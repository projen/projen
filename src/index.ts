export * from "./ai-instructions";
export * from "./compare";
export * from "./component";
export * from "./dev-env";
export * from "./dependencies";
export * from "./docker-compose";
export * from "./dockerfile";
export * from "./file";
export * from "./gitattributes";
export * from "./gitpod";
export * from "./runner-options";
export * from "./ignore-file";
export * from "./ini";
export * from "./json";
export * from "./json-patch";
export * from "./logger";
export * from "./license";
export * from "./makefile";
export * from "./object-file";
export * from "./option-hints";
export * from "./project";
export * from "./project-build";
export * from "./project-tree";
export * from "./projects";
export * from "./projenrc";
export * from "./projenrc-json";
export * from "./readme";
export * from "./renovatebot";
export * from "./sample-file";
export * from "./semver";
export * from "./source-code";
export * from "./task";
export * from "./tasks";
export * from "./task-model";
export * from "./task-runtime";
export * from "./testing";
export * from "./textfile";
export * from "./toml";
export * from "./version";
export * from "./yaml";
export * from "./xmlfile";

// export submodules
export * as web from "./web";
export * as github from "./github";
export * as gitlab from "./gitlab";
export * as circleci from "./circleci";
export * as vscode from "./vscode";
export * as java from "./java";
export * as python from "./python";
export * as typescript from "./typescript";
export * as javascript from "./javascript";
export * as release from "./release";
export * as awscdk from "./awscdk";
export * as cdk8s from "./cdk8s";
export * as cdktf from "./cdktf";
export * as cdk from "./cdk";
export * as build from "./build";
