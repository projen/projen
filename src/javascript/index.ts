export * from "./bundler";
export * from "./eslint";
export * from "./jest";
export * from "./license-checker";
export * from "./node-package";
export * from "./node-project";
export * from "./npm-config";
export * from "./prettier";
export * from "./projenrc";
export * from "./typescript-config";
export * from "./upgrade-dependencies";
export * from "./yarnrc";
