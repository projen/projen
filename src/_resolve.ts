import { types } from "util";
import { ResolveOptions, IResolvable } from "./file";

function isResolvable(obj: any): obj is IResolvable {
  return (obj as IResolvable).toJSON !== undefined;
}

export function resolve(value: any, options: ResolveOptions = {}): any {
  const args = options.args ?? [];
  const omitEmpty = options.omitEmpty ?? false;

  if (value == null) {
    return value;
  }

  if (isResolvable(value)) {
    const resolved = value.toJSON();
    return resolve(resolved, options);
  }

  // Special resolution for few JavaScript built-in types
  // that by default would be stringified as empty objects ('{}')
  // as they are missing a `toJSON` implementation.
  switch (true) {
    case types.isRegExp(value):
      if (value.flags) {
        throw new Error(
          "RegExp with flags should be explicitly converted to a string"
        );
      }
      return value.source;

    case types.isSet(value):
      return resolve(Array.from(value), options);

    case types.isMap(value):
      return resolve(Object.fromEntries(value), options);

    case types.isBoxedPrimitive(value):
      return resolve(value.valueOf(), options);
  }

  // if value is a function, call it and resolve the result.
  if (typeof value === "function") {
    const resolved = value.apply(undefined, args);
    return resolve(resolved, options);
  }

  if (typeof value !== "object") {
    return value;
  }

  if (Array.isArray(value)) {
    if (omitEmpty && value.length === 0) {
      return undefined;
    }
    return value.map((x) => resolve(x, options)).filter((x) => x != null); // filter undefined/null/omitted
  }

  // only allow data types (i.e. objects without constructors)
  if (value.constructor && value.constructor?.name !== "Object") {
    throw new Error(
      `only data types can be resolved. trying to resolve object of type ${value.constructor?.name}`
    );
  }

  const result: any = {};

  for (const [k, v] of Object.entries(value)) {
    const resolved = resolve(v, options);

    // skip undefined values
    if (resolved === undefined) {
      continue;
    }

    result[k] = resolved;
  }

  if (omitEmpty && Object.keys(result).length === 0) {
    return undefined;
  }

  return result;
}
