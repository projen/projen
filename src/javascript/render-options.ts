import * as inventory from "../inventory";
import { InitProjectOptionHints } from "../option-hints";
import { ModuleImports } from "./private/modules";

const PROJEN_NEW = "__new__";
const TAB = makePadding(2);

/**
 * Options for `renderProjectOptions`.
 */
export interface RenderProjectOptions {
  /**
   * The project type to render.
   */
  readonly type: inventory.ProjectType;

  /**
   * Project arguments as passed to `projen new`.
   */
  readonly args: Record<string, any>;

  /**
   * Include commented out options.
   * @default InitProjectOptionHints.FEATURED
   */
  readonly comments?: InitProjectOptionHints;

  /**
   * Inject a `__new__` attribute to the project constructor with a stringified
   * version of the project parameters and a `jsiiFqn` attribute that includes
   * the FQN of the project type. This is needed in order to generate initial
   * projenrc files.
   *
   * @default false
   */
  readonly bootstrap?: boolean;

  /**
   * A list of fields to omit from the initial projenrc file.
   * @default - none
   */
  readonly omitFromBootstrap?: string[];

  /**
   * Prefix all imports with this string and the full module name
   * This is required when executing options code in a vm
   *
   * @default - only use submodule as prefix
   */
  readonly prefixImports?: string;
}

/**
 * Information passed from `projen new` to the project object when the project
 * is first created. It is used to generate projenrc files in various languages.
 */
interface ProjenInit {
  /**
   * The JSII FQN of the project type.
   */
  readonly fqn: string;

  /**
   * Initial arguments passed to `projen new`.
   */
  readonly args: Record<string, any>;

  /**
   * Include commented out options. Does not apply to projenrc.json files.
   */
  readonly comments: InitProjectOptionHints;
}

/**
 * Renders options as if the project was created via `projen new` (embeds the __new__ field).
 */
export function renderProjenInitOptions(
  fqn: string,
  args: Record<string, any>,
  comments: InitProjectOptionHints = InitProjectOptionHints.NONE
): any {
  return {
    ...args,
    [PROJEN_NEW]: { fqn, args, comments } as ProjenInit,
  };
}

export function resolveInitProject(opts: any) {
  const f = opts[PROJEN_NEW] as ProjenInit;
  if (!f) {
    return undefined;
  }

  const type = inventory.resolveProjectType(f.fqn);
  if (!type) {
    throw new Error(`unable to resolve project type for ${f.fqn}`);
  }
  return {
    args: f.args,
    fqn: f.fqn,
    type: type,
    comments: f.comments,
  };
}

/**
 * Prints all parameters that can be used in a project type, alongside their descriptions.
 *
 * Parameters in `params` that aren't undefined are rendered as defaults,
 * while all other parameters are rendered as commented out.
 *
 * Returns the printed output and a set of required imports as an object
 * in the form { options, imports }.
 */
export function renderJavaScriptOptions(opts: RenderProjectOptions): {
  renderedOptions: string;
  imports: ModuleImports;
} {
  const renders: Record<string, string> = {};
  const optionsWithDefaults: string[] = [];
  const allImports = new ModuleImports();

  for (const option of opts.type.options) {
    if (option.deprecated) {
      continue;
    }

    const optionName = option.name;

    if (opts.args[optionName] !== undefined) {
      const arg = opts.args[optionName];
      const { js, moduleName, importName } = renderArgAsJavaScript(arg, option);

      renders[optionName] = `${optionName}: ${js},`;
      optionsWithDefaults.push(optionName);

      if (moduleName && importName) {
        allImports.from(moduleName, importName);
        if (opts.prefixImports) {
          const prefix = `${opts.prefixImports}["${moduleName}"].`;
          renders[optionName] = `${optionName}: ${prefix}${js},`;
        }
      }
    } else {
      const defaultValue = option.default?.startsWith("-")
        ? undefined
        : option.default ?? undefined;
      renders[optionName] = `// ${optionName}: ${defaultValue},`;
    }
  }

  const bootstrap = opts.bootstrap ?? false;
  if (bootstrap) {
    for (const arg of opts.omitFromBootstrap ?? []) {
      delete opts.args[arg];
    }
    renders[PROJEN_NEW] = `${PROJEN_NEW}: ${JSON.stringify({
      args: opts.args,
      fqn: opts.type.fqn,
      comments: opts.comments,
    } as ProjenInit)},`;
    optionsWithDefaults.push(PROJEN_NEW);
  }

  // generate rendering
  const result: string[] = [];
  result.push("{");

  // render options with defaults
  optionsWithDefaults.sort();
  for (const optionName of optionsWithDefaults) {
    result.push(`${TAB}${renders[optionName]}`);
  }
  if (result.length > 1) {
    result.push("");
  }

  // render options without defaults as comments
  if (opts.comments === InitProjectOptionHints.ALL) {
    const options = opts.type.options.filter(
      (opt) => !opt.deprecated && opts.args[opt.name] === undefined
    );
    result.push(...renderCommentedOptionsByModule(renders, options));
  } else if (opts.comments === InitProjectOptionHints.FEATURED) {
    const options = opts.type.options.filter(
      (opt) =>
        !opt.deprecated && opts.args[opt.name] === undefined && opt.featured
    );
    result.push(...renderCommentedOptionsInOrder(renders, options));
  } else if (opts.comments === InitProjectOptionHints.NONE) {
    // don't render any extra options
  }

  if (result[result.length - 1] === "") {
    result.pop();
  }
  result.push("}");
  return { renderedOptions: result.join("\n"), imports: allImports };
}

function renderCommentedOptionsByModule(
  renders: Record<string, string>,
  options: inventory.ProjectOption[]
) {
  const optionsByModule: Record<string, inventory.ProjectOption[]> = {};

  for (const option of options) {
    const parentModule = option.parent;
    optionsByModule[parentModule] = optionsByModule[parentModule] ?? [];
    optionsByModule[parentModule].push(option);
  }

  for (const parentModule in optionsByModule) {
    optionsByModule[parentModule].sort((o1, o2) =>
      o1.name.localeCompare(o2.name)
    );
  }

  const result = [];
  const marginSize = Math.max(
    ...options.map((opt) => renders[opt.name].length)
  );
  for (const [moduleName, optionGroup] of Object.entries(
    optionsByModule
  ).sort()) {
    result.push(`${TAB}/* ${moduleName} */`);
    for (const option of optionGroup) {
      const paramRender = renders[option.name];
      const docstring = option.docs || "No documentation found.";
      result.push(
        `${TAB}${paramRender}${makePadding(
          marginSize - paramRender.length + 2
        )}/* ${docstring} */`
      );
    }
    result.push("");
  }
  return result;
}

function renderCommentedOptionsInOrder(
  renders: Record<string, string>,
  options: inventory.ProjectOption[]
) {
  const result = [];
  const marginSize = Math.max(
    ...options.map((opt) => renders[opt.name].length)
  );
  for (const option of options) {
    const paramRender = renders[option.name];
    const docstring = option.docs || "No documentation found.";
    result.push(
      `${TAB}${paramRender}${makePadding(
        marginSize - paramRender.length + 2
      )}/* ${docstring} */`
    );
  }
  return result;
}

/**
 * Renders a value as a JavaScript value, converting strings to enums where
 * appropriate. The type must be JSON-like (string, number, boolean, array,
 * enum, or JSON object).
 *
 * Returns a JavaScript expression as a string, and the names of any
 * necessary imports.
 */
function renderArgAsJavaScript(arg: any, option: inventory.ProjectOption) {
  if (option.kind === "enum") {
    if (!option.fqn) {
      throw new Error(`fqn field is missing from enum option ${option.name}`);
    }
    const parts = option.fqn.split("."); // -> ['projen', 'web', 'MyEnum']
    const enumChoice = String(arg).toUpperCase().replace(/-/g, "_"); // custom-value -> CUSTOM_VALUE
    const js = `${parts.slice(1).join(".")}.${enumChoice}`; // -> web.MyEnum.CUSTOM_VALUE
    const moduleName = parts[0]; // -> projen
    const importName = parts[1]; // -> web
    return { js, moduleName, importName };
  } else if (option.jsonLike) {
    return { js: JSON.stringify(arg) };
  } else {
    throw new Error(
      `Unexpected option ${option.name} - cannot render a value for this option because it does not have a JSON-like type.`
    );
  }
}

function makePadding(paddingLength: number): string {
  return " ".repeat(paddingLength);
}
