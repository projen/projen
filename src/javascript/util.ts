import { existsSync, readFileSync } from "fs";
import { basename, dirname, extname, join, sep, resolve, posix } from "path";
import * as semver from "semver";
import { NodePackage } from "./node-package";
import { Project } from "../project";
import { findUp } from "../util";

/**
 * Basic interface for `package.json`.
 */
interface PackageManifest {
  [key: string]: any;

  /**
   * Package name.
   */
  name: string;
  /**
   * Package version.
   */
  version: string;
}

export function renderBundleName(entrypoint: string) {
  const parts = join(entrypoint).split(sep);
  if (parts[0] === "src") {
    parts.shift(); // just remove 'src' if its the first element for ergonomics
  }

  const p = parts.join(posix.sep);
  const dir = dirname(p);
  const base = basename(p, extname(p));
  return posix.join(dir, base);
}

/**
 * Regex for AWS CodeArtifact registry
 */
export const codeArtifactRegex =
  /^https:\/\/(?<registry>(?<domain>[^\.]+)-(?<accountId>\d{12})\.d\.codeartifact\.(?<region>[^\.]+).*\.amazonaws\.com\/.*\/(?<repository>[^\/]+)\/)/;

/**
 * gets AWS details from the Code Artifact registry URL
 * throws exception if not matching expected pattern
 * @param registryUrl Code Artifact registry URL
 * @returns object containing the (domain, accountId, region, repository)
 */
export function extractCodeArtifactDetails(registryUrl: string) {
  const match = registryUrl.match(codeArtifactRegex);
  if (match?.groups) {
    const { domain, accountId, region, repository, registry } = match.groups;
    return { domain, accountId, region, repository, registry };
  }
  throw new Error("Could not get CodeArtifact details from npm Registry");
}

export function minVersion(version: string): string | undefined {
  if (semver.validRange(version)) {
    return semver.minVersion(version)?.version;
  } else {
    return version;
  }
}

/**
 * Attempt to resolve location of the given `moduleId`.
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModule(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  try {
    return require.resolve(moduleId, options);
  } catch {
    return undefined;
  }
}

/**
 * Attempt to resolve a module's manifest (package.json) path via `require.resolve` lookup.
 *
 * @remarks
 * If the target package has `exports` that differ from the default
 * (i.e, it defines the `exports` field in its manifest) and does not
 * explicitly include an entry for `package.json`, this strategy will fail.
 * See {@link tryResolveManifestPathFromDefaultExport} as an alternative.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveModuleManifestPath(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  // cannot just `require('dependency/package.json')` here because
  // `options.paths` may not overlap with this node proc's resolution paths.
  const manifestId = `${moduleId}/package.json`;
  return tryResolveModule(manifestId, options);
}

/**
 * Attempt to resolve a module's manifest (package.json) path by looking for the nearest
 * `package.json` file that is an ancestor to the module's default export location.
 *
 * @param moduleId Module ID to lookup.
 * @param options Passed through to `require.resolve`.
 */
export function tryResolveManifestPathFromDefaultExport(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const defaultExportPath = tryResolveModule(moduleId, options);
  if (!defaultExportPath) {
    return undefined;
  }
  const moduleDir = findUp("package.json", defaultExportPath);
  if (!moduleDir) {
    return undefined;
  }
  return join(moduleDir, "package.json");
}

/**
 * Attempt to resolve a module's manifest (package.json) path by checking for its existence under `node_modules` relative to `basePath`.
 *
 * @remarks
 * This strategy can be helpful in the scenario that a module defines
 * custom exports without `package.json` and no default export (i.e, some type definition packages).
 *
 * @param moduleId Module ID to lookup.
 * @param basePath Root path to search from.
 */
export function tryResolveManifestPathFromPath(
  moduleId: string,
  basePath: string
) {
  const base = basePath.includes("node_modules")
    ? basePath
    : join(basePath, "node_modules");
  const filePath = resolve(base, ...moduleId.split("/"), "package.json");
  if (existsSync(filePath)) {
    return filePath;
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) path by searching for it in the optionally provided paths array
 * as well as the current node processes' default resolution paths.
 * @param moduleId Module ID to search for.
 * @param options Search options.
 */
export function tryResolveManifestPathFromSearch(
  moduleId: string,
  options?: { paths: string[] }
): string | undefined {
  const searchPaths = [
    ...(options?.paths ?? []),
    ...(require.resolve.paths(moduleId) ?? []),
  ];
  for (const path of searchPaths) {
    const result = tryResolveManifestPathFromPath(moduleId, path);
    // early return on first result.
    if (result) {
      return result;
    }
  }
  return undefined;
}

/**
 * Attempt to resolve a module's manifest (package.json) using multiple strategies.
 * @param moduleId Module to resolve manifest path for.
 * @param options Resolution options.
 */
export function tryResolveModuleManifest(
  moduleId: string,
  options?: { paths: string[] }
): PackageManifest | undefined {
  const strategies = [
    tryResolveModuleManifestPath,
    tryResolveManifestPathFromDefaultExport,
    tryResolveManifestPathFromSearch,
  ];
  for (const strategy of strategies) {
    const result = strategy(moduleId, options);
    // early return on first result.
    if (result) {
      try {
        const manifest = JSON.parse(
          readFileSync(result, "utf8")
        ) as PackageManifest;
        // verify name matches target module.
        if (manifest.name === moduleId) {
          return manifest;
        }
      } catch {
        // continue to next strategy.
      }
    }
  }
  return undefined;
}

/**
 * Attempt to resolve the installed version of a given dependency.
 * @param dependencyName Name of dependency.
 * @param options Optional options passed through to `require.resolve`.
 */
export function tryResolveDependencyVersion(
  dependencyName: string,
  options?: { paths: string[] }
): string | undefined {
  const manifest = tryResolveModuleManifest(dependencyName, options);
  if (!manifest) {
    return undefined;
  }
  return manifest?.version;
}

/**
 * Whether the given dependency version is installed
 *
 * This can be used to test for the presence of certain versions of devDependencies,
 * and do something dependency-specific in certain Components. For example, test for
 * a version of Jest and generate different configs based on the Jest version.
 *
 * NOTE: The implementation of this function currently is currently
 * approximate: to do it correctly, we would need a separate implementation
 * for every package manager, to query its installed version (either that, or we
 * would code to query `package-lock.json`, `yarn.lock`, etc...).
 *
 * Instead, we will look at `package.json`, and assume that the versions
 * picked by the package manager match ~that. This will work well enough for
 * major version checks, but may fail for point versions.
 *
 * What we SHOULD do is: `actualVersion ∈ checkRange`.
 *
 * What we do instead is a slightly more sophisticated version of
 * `requestedRange ∩ checkRange != ∅`. This will always give a correct result if
 * `requestedRange ⊆ checkRange`, but may give false positives when `checkRange
 * ⊆ requestedRange`.
 *
 * May return `undefined` if the question cannot be answered. These include the
 * following cases:
 *
 *   - The dependency is requested via local file dependencies (`file://...`)
 *   - The dependency uses an other type of URL, such as a GitHub URL
 *   - The dependency is not found in the `package.json`, such as when
 *     bootstrapping from an external projen package, and the `package.json`
 *     file only has that one external package as a dependency
 *
 * Otherwise it will return `true` if the installed version is probably in the
 * requested range, and `false` if it is probably not.
 *
 * This API may eventually be added to the public projen API, but only after
 * we implement exact version checking.
 *
 * @param dependencyName The name of the dependency
 * @param checkRange A particular version, or range of versions.
 */
export function hasDependencyVersion(
  project: Project,
  dependencyName: string,
  checkRange: string
): boolean | undefined {
  const file = NodePackage.of(project)?.file;
  if (!file) {
    return undefined;
  }

  if (!existsSync(file.absolutePath)) {
    return undefined;
  }

  const pj = JSON.parse(readFileSync(file.absolutePath, "utf-8"));

  // Technicaly, we should be intersecting all ranges to come up with the most narrow dependency
  // range, but `semver` doesn't allow doing that and we don't want to add a dependency on `semver-intersect`.
  //
  // Let's take the first dependency declaration we find, and assume that people
  // set up their `package.json` correctly.
  let requestedRange: string | undefined;
  for (const key of ["dependencies", "devDependencies", "peerDependencies"]) {
    const deps = pj[key] ?? {};
    let requestedVersion = deps[dependencyName];
    if (requestedVersion) {
      // If this is not a valid range, it could be 'file:dep.tgz', or a GitHub URL. No way to know what
      // version we're getting, bail out.
      if (!semver.validRange(requestedVersion)) {
        return undefined;
      }
      requestedRange = requestedVersion;
      break;
    }
  }

  // If the dependency is not found in the `package.json`, we can't answer the question (yet).
  // This means that the dependency hasn't been added yet, which means we know (upstream) what we're going to request,
  // or we're going to ask for '*' and we'll get the latest version.
  if (!requestedRange) {
    return undefined;
  }

  return installedVersionProbablyMatches(requestedRange, checkRange);
}

/**
 * Whether the given requestedRange *probably* leads to the installation of a version that matches checkRange
 *
 * We assume that NPM always installs the most recent version of a package that
 * is allowed by the requestedRange.
 */
export function installedVersionProbablyMatches(
  requestedRange: string,
  checkRange: string
): boolean {
  const options = {
    includePrerelease: true,
    loose: true,
  };

  // No questions asked: always true
  if (semver.subset(requestedRange, checkRange, options)) {
    return true;
  }

  // Also no questions asked: always false
  if (!semver.intersects(requestedRange, checkRange, options)) {
    return false;
  }

  // Now we're in tricky territory. We intersect, but aren't a full subset.
  // We are in one of the following 2 situations, which we will tie-break by
  // assuming NPM will install the most recent matching version in 'requested'.
  //
  // requested  | check    | result
  // -----------|----------|-----------
  //   >= 2     |  >= 3    | probably true (chance of FP)
  //   <= 2     |  <= 1    | probably false (change of FN)
  //
  // `semver` doesn't make it easy to distinguish these cases (we can't request
  // the `maxVersion` that satisfies a range). Instead what we do is
  // get the `minVersion` of each range, and if they compare equal we assume
  // we're in the bottom case with `<=` checks, and return `false`.

  return !semver.eq(
    semver.minVersion(requestedRange, options) ?? "1.2.3",
    semver.minVersion(checkRange, options) ?? "1.2.3"
  );
}
