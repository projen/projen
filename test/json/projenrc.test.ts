import * as path from "path";
import { renderProjenInitOptions } from "../../src/javascript/render-options";
import { Projenrc } from "../../src/projenrc-json";
import { synthSnapshot, TestProject, withProjectDir } from "../util";

test("projenrc.json default project", () => {
  // GIVEN
  const project = new TestProject();

  // WHEN
  new Projenrc(project);

  // THEN
  expect(synthSnapshot(project)).toMatchSnapshot();
});

test("projenrc.json with typed options", () => {
  // GIVEN
  const project = new TestProject(
    renderProjenInitOptions("projen.typescript.TypeScriptProject", {
      staleOptions: {
        issues: {
          daysBeforeStale: 100, // number, nested option
        },
        pullRequest: {
          enabled: false, // boolean, nested option
        },
      },
      name: "@example/foo", // string
      npmAccess: "public", // enum provided as string
    })
  );

  // WHEN
  new Projenrc(project);

  // THEN
  expect(synthSnapshot(project)).toMatchSnapshot();
});

test("projenrc.json new project in outdir", () => {
  withProjectDir((projectdir) => {
    const newOutDir = path.join(projectdir, "foo");
    const project = new TestProject(
      renderProjenInitOptions("projen.typescript.TypeScriptProject", {
        outdir: newOutDir,
      })
    );
    const projen = new Projenrc(project);

    expect(projen.project.outdir).toEqual(newOutDir);
  });
});
