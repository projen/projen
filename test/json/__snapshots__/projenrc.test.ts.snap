// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`projenrc.json default project 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.json and run "npx projen".

* text=auto eol=lf
/.gitattributes linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.gitignore linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.json and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: amannn/action-semantic-pull-request@v6
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.json and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/.github/workflows/pull-request-lint.yml
!/.projenrc.json
",
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.json and run "npx projen".",
    "files": [
      ".gitattributes",
      ".github/workflows/pull-request-lint.yml",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.json and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "env": {
          "FILENAME": ".projenrc.json",
        },
        "name": "default",
        "steps": [
          {
            "builtin": "run-projenrc-json",
          },
        ],
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  "README.md": "# replace this",
}
`;

exports[`projenrc.json with typed options 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.json and run "npx projen".

* text=auto eol=lf
/.gitattributes linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.gitignore linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.json and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: amannn/action-semantic-pull-request@v6
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.json and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/.github/workflows/pull-request-lint.yml
!/.projenrc.json
",
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.json and run "npx projen".",
    "files": [
      ".gitattributes",
      ".github/workflows/pull-request-lint.yml",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.json and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
      },
      "default": {
        "description": "Synthesize project files",
        "env": {
          "FILENAME": ".projenrc.json",
        },
        "name": "default",
        "steps": [
          {
            "builtin": "run-projenrc-json",
          },
        ],
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
      },
    },
  },
  ".projenrc.json": {
    "name": "@example/foo",
    "npmAccess": "public",
    "staleOptions": {
      "issues": {
        "daysBeforeStale": 100,
      },
      "pullRequest": {
        "enabled": false,
      },
    },
    "type": "projen.typescript.TypeScriptProject",
  },
  "README.md": "# replace this",
}
`;
